[manager/nuget] packages.lock file should be updated when dotnet-sdk is updated

### Discussed in https://github.com/renovatebot/renovate/discussions/27977

<div type='discussions-op-text'>

<sup>Originally posted by **marcmognol** March 17, 2024</sup>
### Tell us more.

When a new version of dotnet sdk is released, Renovate creates a new PR to update the global.json.
But Renovate doesn't update the packages.lock.json file, and as some nuget packages are updated along the sdk, our CI pipelines fails and I need to update packages.lock json file manually.

I would like Renovate run a `dotnet restore` whenever the global.json is updated.

Thanks</div>

Projects containing an implicit reference to `FSharp.Core` (ie. every F# project) are guaranteed to break with every dotnet monorepo update:

`chore(deps): update dotnet monorepo to v8.0.401`

```
$ dotnet restore --locked-mode
  Determining projects to restore...
/.../Build.fsproj : error NU1004: The package reference FSharp.Core version has changed from [8.0.301, ) to [8.0.400, ).The packages lock file is inconsistent with the project dependencies so restore can't be run in locked mode. Disable the RestoreLockedMode MSBuild property or pass an explicit --force-evaluate option to run restore to update the lock file. [/.../Solution.sln]
```

--- a/lib/modules/manager/nuget/artifacts.spec.ts
+++ b/lib/modules/manager/nuget/artifacts.spec.ts
@@ -2,6 +2,7 @@ import upath from 'upath';
 import { mockDeep } from 'vitest-mock-extended';
 import { GlobalConfig } from '../../../config/global';
 import type { RepoGlobalConfig } from '../../../config/types';
+import { TEMPORARY_ERROR } from '../../../constants/error-messages';
 import * as docker from '../../../util/exec/docker';
 import type { UpdateArtifactsConfig } from '../types';
 import * as util from './util';
@@ -43,6 +44,25 @@ describe('modules/manager/nuget/artifacts', () => {
 
   afterEach(() => {
     GlobalConfig.reset();
+  });
+
+  it('re-throws TEMPORARY_ERROR', async () => {
+    fs.getSiblingFileName.mockReturnValueOnce('packages.lock.json');
+    git.getFiles.mockResolvedValueOnce({
+      'packages.lock.json': 'Current packages.lock.json',
+    });
+    fs.getLocalFiles.mockResolvedValueOnce({
+      'packages.lock.json': 'New packages.lock.json',
+    });
+    fs.writeLocalFile.mockRejectedValueOnce(new Error(TEMPORARY_ERROR));
+    await expect(
+      nuget.updateArtifacts({
+        packageFileName: 'project.csproj',
+        updatedDeps: [{ depName: 'dep' }],
+        newPackageFileContent: '{}',
+        config,
+      }),
+    ).rejects.toThrow(TEMPORARY_ERROR);
   });
 
   it('aborts if no lock file found', async () => {

--- a/lib/modules/manager/nuget/artifacts.ts
+++ b/lib/modules/manager/nuget/artifacts.ts
@@ -21,6 +21,7 @@ import type {
 } from '../types';
 import { createNuGetConfigXml } from './config-formatter';
 import {
+  GLOBAL_JSON,
   MSBUILD_CENTRAL_FILE,
   NUGET_CENTRAL_FILE,
   getDependentPackageFiles,
@@ -98,8 +99,11 @@ export async function updateArtifacts({
     packageFileName.endsWith(`/${NUGET_CENTRAL_FILE}`) ||
     packageFileName.endsWith(`/${MSBUILD_CENTRAL_FILE}`);
 
+  const isGlobalJson = packageFileName === GLOBAL_JSON;
+
   if (
     !isCentralManagement &&
+    !isGlobalJson &&
     !regEx(/(?:cs|vb|fs)proj$/i).test(packageFileName)
   ) {
     // This could be implemented in the future if necessary.
@@ -116,6 +120,7 @@ export async function updateArtifacts({
   const deps = await getDependentPackageFiles(
     packageFileName,
     isCentralManagement,
+    isGlobalJson,
   );
   const packageFiles = deps.filter((d) => d.isLeaf).map((d) => d.name);
 
@@ -176,7 +181,6 @@ export async function updateArtifacts({
 
     return retArray.length > 0 ? retArray : null;
   } catch (err) {
-    // istanbul ignore if
     if (err.message === TEMPORARY_ERROR) {
       throw err;
     }

--- a/lib/modules/manager/nuget/extract.spec.ts
+++ b/lib/modules/manager/nuget/extract.spec.ts
@@ -6,6 +6,7 @@ import { DotnetVersionDatasource } from '../../datasource/dotnet-version';
 import type { ExtractConfig } from '../types';
 import { extractPackageFile } from '.';
 import { Fixtures } from '~test/fixtures';
+import { fs } from '~test/util';
 
 const config: ExtractConfig = {};
 
@@ -41,8 +42,10 @@ describe('modules/manager/nuget/extract', () => {
     it('extracts package file version', async () => {
       const packageFile = 'sample.csproj';
       const sample = Fixtures.get(packageFile);
+      vi.spyOn(fs, 'localPathExists').mockResolvedValueOnce(true);
       const res = await extractPackageFile(sample, packageFile, config);
       expect(res?.packageFileVersion).toBe('0.1.0');
+      expect(res?.lockFiles).toEqual(['packages.lock.json']);
     });
 
     it('does not fail on package file without version', async () => {

--- a/lib/modules/manager/nuget/extract.ts
+++ b/lib/modules/manager/nuget/extract.ts
@@ -216,7 +216,6 @@ export async function extractPackageFile(
 
   const res: PackageFileContent = { deps, packageFileVersion };
   const lockFileName = getSiblingFileName(packageFile, 'packages.lock.json');
-  // istanbul ignore if
   if (await localPathExists(lockFileName)) {
     res.lockFiles = [lockFileName];
   }

--- a/lib/modules/manager/nuget/package-tree.spec.ts
+++ b/lib/modules/manager/nuget/package-tree.spec.ts
@@ -94,6 +94,26 @@ describe('modules/manager/nuget/package-tree', () => {
       ]);
     });
 
+    it('returns project for two projects with one reference and global.json', async () => {
+      scm.getFileList.mockResolvedValue(['one/one.csproj', 'two/two.csproj']);
+      Fixtures.mock({
+        '/tmp/repo/one/one.csproj': Fixtures.get(
+          'two-one-reference-with-central-versions/one/one.csproj',
+        ),
+        '/tmp/repo/two/two.csproj': Fixtures.get(
+          'two-one-reference-with-central-versions/two/two.csproj',
+        ),
+        '/tmp/repo/global.json': '{}',
+      });
+
+      expect(
+        await getDependentPackageFiles('global.json', false, true),
+      ).toEqual([
+        { isLeaf: false, name: 'one/one.csproj' },
+        { isLeaf: true, name: 'two/two.csproj' },
+      ]);
+    });
+
     it('returns projects for three projects with two linear references', async () => {
       scm.getFileList.mockResolvedValue([
         'one/one.csproj',

--- a/lib/modules/manager/nuget/package-tree.ts
+++ b/lib/modules/manager/nuget/package-tree.ts
@@ -7,6 +7,7 @@ import { scm } from '../../platform/scm';
 import type { ProjectFile } from './types';
 import { readFileAsXmlDocument } from './util';
 
+export const GLOBAL_JSON = 'global.json';
 export const NUGET_CENTRAL_FILE = 'Directory.Packages.props';
 export const MSBUILD_CENTRAL_FILE = 'Packages.props';
 
@@ -16,6 +17,7 @@ export const MSBUILD_CENTRAL_FILE = 'Packages.props';
 export async function getDependentPackageFiles(
   packageFileName: string,
   isCentralManagement = false,
+  isGlobalJson = false,
 ): Promise<ProjectFile[]> {
   const packageFiles = await getAllPackageFiles();
   const graph = new Graph();
@@ -24,16 +26,24 @@ export async function getDependentPackageFiles(
     graph.addNode(packageFileName);
   }
 
+  if (isGlobalJson) {
+    graph.addNode(GLOBAL_JSON);
+  }
+
   const parentDir =
     packageFileName === NUGET_CENTRAL_FILE ||
-    packageFileName === MSBUILD_CENTRAL_FILE
+    packageFileName === MSBUILD_CENTRAL_FILE ||
+    packageFileName === GLOBAL_JSON
       ? ''
       : upath.dirname(packageFileName);
 
   for (const f of packageFiles) {
     graph.addNode(f);
 
-    if (isCentralManagement && upath.dirname(f).startsWith(parentDir)) {
+    if (
+      (isCentralManagement || isGlobalJson) &&
+      upath.dirname(f).startsWith(parentDir)
+    ) {
       graph.addEdge(packageFileName, f);
     }
   }
@@ -70,7 +80,7 @@ export async function getDependentPackageFiles(
   const deps = new Map<string, boolean>();
   recursivelyGetDependentPackageFiles(packageFileName, graph, deps);
 
-  if (isCentralManagement) {
+  if (isCentralManagement || isGlobalJson) {
     // remove props file, as we don't need it
     deps.delete(packageFileName);
   }

